File: MatlabCodes/Fig06_01.m
function COS_Normal_Density_Recovery
close all;clc;
format long
i = complex(0,1); %assigning i=sqrt(-1)

% The COS method settings

a = -10; % truncation domain
b =  10;

% Characteristic function parameters

mu     = 0;    sigma  = 1;

% Domain for the density f(x)

x = [-5, -4, -3, -2, -1, 0, 1, 2, 3, 4, 5];

% Characteristic function  

cf  =  @(u) exp(i * mu * u - 0.5 * sigma^2 * u.^2);

% Iterate over different numbers of expansion terms

Ngrid = 2.^(2:1:6);
idx = 1;
error = zeros(length(Ngrid),1);
figure(1); hold on; grid on;xlabel('x');ylabel('f_X(x)')
for N=Ngrid

    % Density from the COS method

    f_X = COSDensity(cf,x,N,a,b);

    % Error

    error(idx) = max(normpdf(x)-f_X);
    idx = idx + 1;
    plot(x,f_X)
end

error

function f_X = COSDensity(cf,x,N,a,b)
i = complex(0,1); %assigning i=sqrt(-1)
k = 0:N-1; 
u = k * pi / (b - a);

% F_k coefficients

F_k    = 2 / (b - a) * real(cf(u) .* exp(-i * u * a));
F_k(1) = F_k(1) * 0.5; % adjustment for the first term

% Final calculation

f_X = F_k * cos(u' * (x - a));

File: PythonCodes/Chapter 06/Fig06_01.py
#%%
"""
Normal density recovery using the COS method
"""
import numpy as np
import matplotlib.pyplot as plt
import scipy.stats as st

def COSDensity(cf,x,N,a,b):
    i = np.complex(0.0,1.0) #assigning i=sqrt(-1)
    k = np.linspace(0,N-1,N)
    u = np.zeros([1,N])
    u = k * np.pi / (b-a)
        
    # F_k coefficients

    F_k    = 2.0 / (b - a) * np.real(cf(u) * np.exp(-i * u * a));
    F_k[0] = F_k[0] * 0.5; # adjustment for the first term
    
    # Final calculation

    f_X = np.matmul(F_k , np.cos(np.outer(u, x - a )))
        
    # We output only the first row

    return f_X
    
def mainCalculation():
    i = np.complex(0.0, 1.0) #assigning i=sqrt(-1)
    
    # Setting for the COS method 

    a = -10.0
    b = 10.0
    
    # Define the range for the expansion points

    N = [2**x for x in range(2,7,1)]
    
    # Setting for normal distribution

    mu = 0.0
    sigma = 1.0 
        
    # Define characteristic function for the normal distribution

    cF = lambda u : np.exp(i * mu * u - 0.5 * np.power(sigma,2.0) * np.power(u,2.0));
    
    # Define the domain for density

    x = np.linspace(-5.0,5,11)
    f_XExact = st.norm.pdf(x,mu,sigma)
    
    plt.figure(1)
    plt.grid()
    plt.xlabel("x")
    plt.ylabel("$f_X(x)$")
    for n in N:
        f_X = COSDensity(cF,x,n,a,b)
        error = np.max(np.abs(f_X-f_XExact))
        print("For {0} expanansion terms the error is {1}".format(n,error))
        
        plt.plot(x,f_X)
    
    
mainCalculation()
